### PR TITLE
Allow JWKSet with single key without kid set.

### LINKED ISSUE #139: OpenIdConnect JWT "kid" is not optional for single key keysets.
Hey!

### Issue

I'm implementing a OpenIDConnect Provider for [Unity IDM](https://www.unity-idm.eu/) and encountered the problem, that the JWT `kid` header attribute is required, even if the `jwks_uri` only returns a single key.

An `InvalidJWT('No kid inside header, but $privateKeyOrSecret specified as JWKSet')`-Exception is thrown.

The [`OpenIDConnect/AbstractProvider::getJWKSet`](https://github.com/SocialConnect/auth/blob/2c13ecf9dd9c32a95e968272904b37a34347fcb2/src/OpenIDConnect/AbstractProvider.php#L39-L61)-Method always returns a `JWKSet`, even though the `JWT::decode` Method only expects a `JWKSet` if there are multiple keys present and a string of the key if there is only one key? (That's the reason i report this against this repo and not SocialConnect/jwx)

### My env

PHP: 7.4.1

Library: SocialConnect/auth

Provider: Provider extending `\SocialConnect\OpenIDConnect\AbstractProvider`

Provider options:

```php
[
    'options' => []
];
```

Thanks :smiley_cat:

### PR DESCRIPTION
**Pull request by MaPePeR**
Fixes SocialConnect/auth#139.

Sadly I couldn't find any tests for `verifySignature` to extend for this case, so this is untested.

### CODE REVIEW

**Comment by ghost** [comment]
any idea when this could be merged ?

**Comment by MaPePeR** [comment]
From my point of view the PR is finished and "ready to be merged" until further feedback. Maybe @ovr or @ADmad know more.

**Comment by ADmad** [comment]
I checked another popular JWT handling lib `firebase/php-jwt` and it too requires the `kid` to be specified regardless of how many keys the JWKS contains. https://github.com/firebase/php-jwt/blob/edda0f9ee45b8367699804f792a9be6d5175e816/src/JWT.php#L397-L407

Even though a JWKS might only have 1 key, the key could change in future and hence your JWT header must always have the `kid`. So that when the key in JWSK is changed the existing JWT using the older `kid` get invalidated and proper error message can be provided on auth failure.

So I don't think this change should be merged.

**Comment by MaPePeR** [comment]
> I checked another popular JWT handling lib `firebase/php-jwt` and it too requires the `kid` to be specified regardless of how many keys the JWKS contains. https://github.com/firebase/php-jwt/blob/edda0f9ee45b8367699804f792a9be6d5175e816/src/JWT.php#L397-L407

That check is only done if there is an array of keys and not a single key. [See the first 3 lines in the method](https://github.com/firebase/php-jwt/blob/edda0f9ee45b8367699804f792a9be6d5175e816/src/JWT.php#L393-L395).

The [JWK specification says](https://datatracker.ietf.org/doc/html/rfc7517#section-4.5):
> The "kid" value is a case-sensitive string.  Use of this member is OPTIONAL.

The [JWT specification](https://datatracker.ietf.org/doc/html/rfc7519) has no mention of `kid` at all.

> Even though a JWKS might only have 1 key, the key could change in future and hence your JWT header must always have the `kid`. So that when the key in JWSK is changed the existing JWT using the older `kid` get invalidated and proper error message can be provided on auth failure.
> 
> So I don't think this change should be merged.

I agree, that, not having the `kid` set, is bad, because its needed to properly change a key. But that is a problem/decision of the identity provider, that we probably cannot fix in all identity providers out there.
If this would be an identity provider library that exposes a keyset and my proposal was to not set the `kid` when there is only a single key: I'd totally agree with you. That would be stupid.
But: That's not the case. This is a library that tries to interface with other (possibly non-ideal) implementations and configurations out there and therefore should not refuse to do something that is valid according to the specification.

File: src/JWKSet.php [hunk 1]
```diff
@@ -42,6 +42,29 @@ public function findKeyByKid(string $kid)
...
+     */
+    public function getDefaultKey() : JWK
+    {
+        if (!$this->hasDefaultKey()) {
+            throw new \RuntimeException("No default key");
+        }
+        foreach ($this->keys as $jwk) {
+            return new JWK($jwk); //Return first JWK in Keys.
+        }
+        throw new \RuntimeException("Unreachable code reached");
```
Comment (ADmad) [inline src/JWKSet.php]:
```suggestion
        reset($this->keys);

        return new JWK(current($this->keys));
```

File: src/JWKSet.php [hunk 2]
```diff
@@ -42,6 +42,29 @@ public function findKeyByKid(string $kid)
 
         throw new RuntimeException('Unknown key');
     }
+    
+    /**
+     * @return bool
+     */
+    public function hasDefaultKey() : bool
+    {
+        return $this->count() == 1;
```
Comment (ADmad) [inline src/JWKSet.php]:
```suggestion
        return $this->count() === 1;
```

File: src/JWT.php [hunk 3]
```diff
@@ -179,8 +179,8 @@ protected function validateHeader($publicKeyOrSecret, DecodeOptions $options)
                 throw new RuntimeException('Please specify jwk set in DecodeOptions, because there is kid inside header');
             }
         } else {
-            if ($publicKeyOrSecret instanceof JWKSet) {
-                throw new InvalidJWT('No kid inside header, but $privateKeyOrSecret specified as JWKSet');
+            if ($publicKeyOrSecret instanceof JWKSet && !$publicKeyOrSecret->hasDefaultKey()) {
+                throw new InvalidJWT('No kid inside header, but $privateKeyOrSecret specified as JWKSet without default');
```
Comment (ADmad) [inline src/JWT.php]:
```suggestion
                throw new InvalidJWT('No kid inside header, but $publicKeyOrSecret specified as JWKSet without default');
```

**Comment by ADmad** [comment]
> But that is a problem/decision of the identity provider, that we probably cannot fix in all identity providers out there.

Fair enough, we can't control what various providers do.

**Comment by ADmad** [comment]
@ovr The testsuite CI is not hooked to show the results for commit/PRs?

**Comment by ovr** [comment]
> The testsuite CI is not hooked to show the results for commit/PRs?

You are right! I forget to connect this repo to GitLab CI as an external repo. I migrated this repo to GitHub actions. We need to rebase this PR to affect changes. Can you please rebase your PR @MaPePeR?

Thanks

**Comment by ADmad** [comment]
@ovr You need to allow the workflow to run. I don't have write perms on this repo :slightly_smiling_face:.

**Comment by ovr** [comment]
@ADmad Done, I've added you as a collaborator to this repository to allow that too. I'm sorry for late reply.

**Comment by ADmad** [comment]
@ovr Should I do a new minor release?

**Comment by ovr** [comment]
@ADmad yeah, minor or path. It's really hard to classify this change.

**Comment by ADmad** [comment]
It's a feature addition IMO and since the public API has changed a new minor `1.3.0` would be better.

**Pull request merged**